### ISSUE TITLE
parsing json from Gephi 

### ISSUE THREAD

**Issue by Chroedde** [almende/vis#1189]
Hello together,

awesome software. It really matches all my needs.
In principle. But I dont get it to load the edge labels via json parser.

The json file looks like:
{"edges":[{"label":"Label 1","source":"1","target":"2","id":"10","attributes":"{"title":"Title 1","Weight":"1.0"},"color":"rgb(74,168,208)","size":1.0},{"label":"Label 2" ...
with hundreds of labels. The nodes are looking quite similar.

My problem is that the edge labels don't appear. If I look at the edge params with JSONstringify(edges.get(params.edges)) then only the "from", "to", "id", "title" and "weight" values appear.

But where is the rest?

If I creat my own edges with
var edges = [{from: 1, to: 2, label: 'bla bla'}, (...) ];
then everything works fine. The problem seems to be the parsing of my json file.

Does anyone have any help for me?

Thanks a lot!
BR
Christoph

**Comment by AlexDM0** [comment]
Hi,

This is a question on basic javascript, it is not using any functions of vis. The problem is (I guess?) in your JSON somewhere. Without the JSON we cannot do anything but we can't really turn the vis issues section into a javascript helpdesk.

Try http://www.jsoneditoronline.org/ to check if your JSON is fine. If you have vis specific issue we'd be happy to help.

Regards

**Comment by AlexDM0** [comment]
> {"edges":[{"label":"Label 1","source":"1","target":"2","id":"10","attributes":"{"title":"Title 1","Weight":"1.0"},"color":"rgb(74,168,208)","size":1.0},{"label":"Label 2" ...

you see the difference between what you have here and what vis expects? source should be from, attributes is not something vis supports. I'm frankly surprised that it does anything at all.

**Comment by AlexDM0** [comment]
Hi,

Let's continue the discussion here, I'll close the new issue you made for this topic. As I said here, the format of the JSON file you show is not in the form vis expects it. 

If you look at our documentation here 
http://visjs.org/docs/network/#importGephi

We use the vis data transforming method to transform Gephi to vis format. I'll append this in the description in that example. 

This conversion is not complete (not everything maybe converted), we don't use Gephi ourselves so we base our converter on a few sample files we get from the community. 

If you can provide a JSON file we could see if we can improve the converter. 

I'll close your other issue as well as that is also about Gephi to vis. 

Regards

**Comment by Chroedde** [comment]
Ok, I send you my script. It works pretty good now! But however it actually shouldn't? :-)
But there is still my question with the directed/undirected lines.

Thank you a lot an best regards from Germany
Christoph

```javascript
<!DOCTYPE html>
<html>
<head>
	<meta http-equiv="content-type" content="text/html; charset=UTF8">
	<title></title>

	<script src="https://ajax.googleapis.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script>
	<script type="text/javascript" src="./vis.js"></script>
	<link type="text/css" rel="stylesheet" href="./vis.css">

	<style type="text/css">
	body {
		background-color: #fff;
		margin: 10px;
		padding: 0;
		font-family:Tahoma;
	}

	#header {
		font-family: Tahoma;
		font-size: 18pt;
		padding: 10px;
	}
	
	#mynetwork {
		width: 100%;
		height: 750px;
		border: 1px solid lightgray;
		z-index:1;
	}

	#nodeContent {
		position: absolute;
		top: 80px;
		right: 19px;
		width: 450px;
		height: 600px;
		margin: 0;
		padding: 20px;
		z-index: 2;
		border: 1px solid #555;
		background: #e4f0ff;
		font-family: Tahoma;
		overflow-y: scroll;
		font-size: 10pt;
	}
	</style>
</head>

<body style="padding="0" margin="0">

<div id="header">Ueberschrift</div>
<div id="mynetwork"></div>
<div id="nodeContent"></div>

<script type="text/javascript">
function loadJSON(path, success, error) {
  var xhr = new XMLHttpRequest();
  xhr.onreadystatechange = function () {
    if (xhr.readyState === 4) {
      if (xhr.status === 200) {
        success(JSON.parse(xhr.responseText));
      }
      else {
        error(xhr);
      }
    }
  };
  xhr.open('GET', path, true);
  xhr.send();
}

var gephiImported;
loadJSON('./netzwerk.json', redrawAll, function(err) {console.log('error')});

var container = document.getElementById('mynetwork');

var nodes = new vis.DataSet();
var edges = new vis.DataSet();
var data = { nodes: nodes, edges: edges	};

var options = {
	nodes: {
		shape: 'dot',
		scaling: {
			min: 30,
			max: 50,
			label: {
				min: 20,
				max: 40,
				drawThreshold: 5,
				maxVisible: 40
			}
		},
		font: { face: 'Tahoma' },
	},
	edges: {
		width: 2,
		font: {
			color: '#777777',
			size: 14, // px
			face: 'Tahoma',
			background: 'none',
			strokeWidth: 0, // px
			align:'top'
		},
		hidden: false,
		hoverWidth: 1.5,
		label: undefined,		// vis.js: Zeile 41821
		labelHighlightBold: true,
		length: undefined,
		selfReferenceSize: 20,
		smooth: { type: 'continuous' },
		selectionWidth: function (width) { return width*5; },
		title: undefined		// vis.js: Zeile 41823
	},
	interaction: {
		dragNodes: true,
		hover: true,
		multiselect: true,
		navigationButtons: true,
		tooltipDelay: 200,
		hideEdgesOnDrag: false
	},
	physics: false
	};

var network;
network = new vis.Network(container, data, options);

network.on('click', function (params) {		

	/* Detailinformationen in Info-Div */
	
	if (params.nodes.length > 0) {
		var data = nodes.get(params.nodes[0]);
		var contentfile = "./content/" + JSON.parse(JSON.stringify(data.attributes.contentfile)) + ".html";
		$('#nodeContent').load(contentfile);
	}
	
	/* Node fokussieren */
	
	var nodeId = params.nodes;
	focusNode(nodeId);
})

function focusNode(node_nr) {
	var nodeId = node_nr;
	var options = {
		// position: {x:positionx,y:positiony}, // this is not relevant when focusing on nodes
		scale: 1.0,
		offset: {x:-200,y:0},
		animation: {
			duration: 500,
			easingFunction: 'easeInOutQuad'
		}
	};
	network.focus(nodeId, options);
}

function redrawAll(gephiJSON) {
	if (gephiJSON.nodes === undefined) {
		gephiJSON = gephiImported;
	}
	else {
		gephiImported = gephiJSON;
	}

	nodes.clear();
	edges.clear();

	var parsed = vis.network.gephiParser.parseGephi(gephiJSON, {
		//fixed: false,
		//parseColor: false
	}); 

	// add the parsed data to the DataSets.
	nodes.add(parsed.nodes);
	edges.add(parsed.edges);

	var contentfile = "./content/start.html";
	$('#nodeContent').load(contentfile);
	
    network.fit(); // zoom to fit
}
</script>

</body>
</html>
```

**Comment by Chroedde** [comment]
The json file is

```javascript
{
  "edges": [
    {
      "label": "Edge 1",
      "source": "1",
      "target": "2",
      "id": "1",
      "attributes": {
        "title": "Title Edge 1",
        "Weight": "1.0"
      },
      "color": "rgb(74,168,208)",
      "size": 1
    }
  ],
  "nodes": [
    {
      "label": "Node 1",
      "x": -545.59228515625,
      "y": 742.2496337890625,
      "id": "1",
      "attributes": {
        "Modularity Class": "2",
        "Number of triangles": "0",
        "contentfile": "node_1",
        "Degree": "3",
        "PageRank": "0.0518578989623337",
        "Clustering Coefficient": "0.0",
        "Weighted Degree": "3.0"
      },
      "color": "rgb(253,196,160)",
      "size": 8.333333015441895
    },
    {
      "label": "Node 2",
      "x": 113.09465789794922,
      "y": 1198.474853515625,
      "id": "2",
      "attributes": {
        "Modularity Class": "4",
        "Number of triangles": "0",
        "contentfile": "node_2",
        "Degree": "10",
        "PageRank": "0.16873345574200924",
        "Clustering Coefficient": "0.0",
        "Weighted Degree": "10.0"
      },
      "color": "rgb(179,0,0)",
      "size": 20
    }
  ]
}
```

**Comment by Chroedde** [comment]
And as I said I inserted the lines
	  edge['label'] = gEdge.label;
and
	  edge['title'] = gEdge.attributes.title;
at line 41821 in vis.js

**Comment by AlexDM0** [comment]
Hi, 

I see, I've misunderstood the initial issue. I have added support for labels and titles in both edges and nodes when importing from gephi.

Could you give me a json with the directed/undirected edges? I'll implement the arrows as well.

I'll add these changes in the next release.

Regards

**Comment by Chroedde** [comment]
That sounds great!! Thank you very much!

Is there still a problem with the difference in "source" to "from" and "target" to "to"?

Regarding the directed/undirected edges: Gephi has a column labeled "Type" for every edge. Here you can decide if an edge is directed (i.e. an arrow) or undirected. Unfortunately, the json export module doesn't export this column so that I have to do it by hand. The json then looks like this:

```javascript
 "edges": [
    {
      "label": "Edge 1",
      "type": "Undirected",  // or "Directed"
      "source": "1",
      "target": "2",
      "id": "1",
      "attributes": {
        "title": "Title Edge 1",
        "Weight": "1.0"
      },
      "color": "rgb(74,168,208)",
      "size": 1
    }
  ]
```
vis.js of course doesn't know the difference between undirected and directed. Would it be hard do make it possible? Background is that I have a network where some connections are mutual and some are cleary directed.

**Comment by AlexDM0** [comment]
Hi,

the problem of the source and target is solved by the gephi converter, don't worry about that. I was not clear to me that you were using the vis gephi parser from your initial post, which is mostly due to my lack of experience with gephi.

Vis has the arrows, and they can be in multiple forms.
http://visjs.org/examples/network/edgeStyles/arrows.html

I assume directed would go from source --> target?

Regards

**Comment by Chroedde** [comment]
Yes. But as also undirected connections have "from" and "to" values they also are displayed as arrows if arrows are switched on in vis, even if the type is undirected... But I think it isn't worth it for you to think about a solution as long as the type column isn't exported...

**Comment by AlexDM0** [comment]
Well no, but the vis edges can be configured to be from from to to, from to to from, in the middle from from
to to or both directions. 

If I assume that directed is always from -> to I won't have to take other directions into account. 

Regards

**Comment by AlexDM0** [comment]
I have added this to develop

**Comment by AlexDM0** [comment]
added in 4.8
